src/core/FailoverJsonRpcProvider.ts

```diff
@@ -0,0 +1,164 @@
+      throw new Error('At least one RPC endpoint is required.');
+    }
+
+    // Validate all endpoints.
+    rpcEndpoints.forEach((endpoint, index) => {
+      if (typeof endpoint === 'string') {
+        try {
+          new URL(endpoint); // eslint-disable-line no-new
+        } catch (e) {
+          throw new Error(`Invalid URL at index ${index}: ${endpoint}.`);
+        }
+      } else if (!(endpoint instanceof FetchRequest)) {
+        throw new Error(`Invalid endpoint type at index ${index}.`);
+      }
+    });
+
+    super(rpcEndpoints[0], network, options);
+    this._rpcEndpoints = rpcEndpoints;
+    this._logger = config.logger;
+
+    this._logger?.info(
+      `Initializing FailoverJsonRpcProvider:\n` +
+        `  Primary endpoint: ${this._getRpcEndpointUrl(rpcEndpoints[0])}\n` +
+        `  Backup endpoints: ${
+          rpcEndpoints
+            .slice(1)
+            .map((e) => this._getRpcEndpointUrl(e))
+            .join(', ') || 'none'
+        }\n`,
+    );
+  }
+
+  /**
+   * Overrides the `_send` method to try each endpoint until the request succeeds.
+   *
+   * While the actual request sending mechanism remains the same as the base class,
+   * we create a new `FetchRequest` for each endpoint instead of using `_getConnection()`,
+   * which would always return the primary (first) endpoint.
+   *
+   * @param payload - The JSON-RPC payload or array of payloads to send.
+   * @returns A promise that resolves to the result of the first successful JSON-RPC call.
+   */
+  public override async _send(
+    payload: JsonRpcPayload | Array<JsonRpcPayload>,
+  ): Promise<Array<JsonRpcResult>> {
+    const errors: { rpcEndpoint: string; error: Error }[] = [];
+
+    // Try each endpoint, in order.
+    for (const rpcEndpoint of this._rpcEndpoints) {
+      const endpointUrl = this._getRpcEndpointUrl(rpcEndpoint);
+      const startTime = Date.now();
+
+      try {
+        let request: FetchRequest;
+        if (typeof rpcEndpoint === 'string') {
+          request = new FetchRequest(rpcEndpoint);
+        } else {
+          request = rpcEndpoint.clone();
+        }
+
+        request.body = JSON.stringify(payload);
+        request.setHeader('content-type', 'application/json');
+
+        const response = await request.send();
+
+        if (!response.ok) {
+          throw new Error(
+            `HTTP request failed with status ${response.statusCode}.`,
+          );
+        }
+
+        let resp = response.bodyJson;
+
+        if (!Array.isArray(resp)) {
+          resp = [resp];
+        }
+
+        // Check for RPC-level errors in the response.
+        for (const item of resp) {
+          if (item.error) {
+            throw new Error(`RPC error: ${JSON.stringify(item.error)}.`);
+          }
+        }
+
+        return resp;
+      } catch (error) {
+        const duration = Date.now() - startTime;
+        const normalizedError =
+          error instanceof Error ? error : new Error('Unknown error');
```

Comment:
instead of normalizing to `Unknown error`, maybe we should just log the full thrown object if it's not an `Error` instance. I'm just thinking if something really weird happens it'd be a shame if this would swallow some valuable info